raiden/network/proxies/token_network.py

```diff
@@ -1176,62 +1156,158 @@ def update_transfer(
-        if not transaction_executed or receipt_or_none:
-            if transaction_executed:
-                block = receipt_or_none['blockNumber']
-                to_compare_block = block
-            else:
-                block = checking_block
-                to_compare_block = self.client.block_number()
 
+            if receipt_or_none:
+                # Because the gas estimation succeeded it is known that:
+                # - The channel exists/existed.
+                # - The channel was at the state closed.
+                # - The partner node was the closing address.
+                # - The account had enough balance to pay for the gas when estimate
+                #   gas was called (however there is a race condition for multiple
+                #   transactions #3890)
+
+                # These checks do not have problems with race conditions because
+                # `poll`ing waits for the transaction to be confirmed.
+                mining_block = receipt_or_none['blockNumber']
+
+                if receipt_or_none['cumulativeGasUsed'] == gas_limit:
+                    msg = (
+                        'update transfer failed and all gas was used. Estimate gas '
+                        'may have underestimated update transfer, or succeeded even '
+                        'though an assert is triggered, or the smart contract code '
+                        'has an conditional assert.'
+                    )
+                    raise RaidenUnrecoverableError(msg)
+
+                channel_data = self._detail_channel(
+                    participant1=self.node_address,
+                    participant2=partner,
+                    block_identifier=mining_block,
+                    channel_identifier=channel_identifier,
+                )
+
+                if channel_data.channel_identifier != channel_identifier:
+                    # This is not a problem, the channel identifier can be set
+                    # to 0 if the channel is settled, or it could have a higher
+                    # value if a new channel was opened
+                    msg = (
+                        f'The provided channel identifier does not match the value '
+                        f'on-chain at the block the update transfer was mined ({mining_block}). '
+                        f'provided_channel_identifier={channel_identifier}, '
+                        f'onchain_channel_identifier={channel_data.channel_identifier}'
+                    )
+                    raise RaidenRecoverableError(msg)
+
+                if channel_data.state > ChannelState.CLOSED:
+                    # This should never happen if the settlement window and gas
+                    # price estimation is done properly.
+                    #
+                    # This is a race condition that cannot be prevented,
+                    # therefore it is a recoverable error.
+                    raise RaidenRecoverableError(
```

Comment:
In the previous code this was an unrecoverable, I changed it to move it more torwards https://github.com/raiden-network/raiden/issues/3645